### PR TITLE
[Merged by Bors] - feat: Finitely Presented Modules

### PR DESCRIPTION
**Pull request by erdOne**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Module/FinitePresentation.lean [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+import Mathlib.LinearAlgebra.Isomorphisms
+import Mathlib.LinearAlgebra.FreeModule.Finite.Basic
+/-!
+
+# Finitely Presented Modules
+
+## Main definition
+
+- `Module.FinitePresentation`: A module is finitely presented if it is finitely generated by some
+  set `s` and the kernel of the presentation `Rˢ → M` is also finitely generated.
```
Comment (kbuzzard) [inline Mathlib/Algebra/Module/FinitePresentation.lean:17]:
```suggestion
- `Module.FinitePresentation`: A module is finitely presented if it is generated by some
  finite set `s` and the kernel of the presentation `Rˢ → M` is also finitely generated.
```

File: Mathlib/Algebra/Module/FinitePresentation.lean [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+Suppose `S` is an `R`-algebra, `M` is an `S`-module. Then
+1. If `S` is f.p., then `M` is `R`-f.p. implies `M` is `S`-f.p.
+2. If `S` is both f.p. (as an algebra) and finite (as a module),
+  then `M` is `S`-fp implies that `M` is `R`-f.p.
+3. If `S` is f.p. as a module, then `S` is f.p. as an algebra.
+In particular,
+4. `S` is f.p. as an `R`-module iff it is f.p. as an algebra and is finite as a module.
+
+
+For finitely presented algeras, see `RingTheory/FinitePresentation`.
```
Comment (kbuzzard) [inline Mathlib/Algebra/Module/FinitePresentation.lean:47]:
```suggestion
For finitely presented algebras, see `RingHom.FinitePresentation` in file `RingTheory/FinitePresentation`.
```

File: Mathlib/Algebra/Module/FinitePresentation.lean [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+    Module.FinitePresentation R M ↔ Module.Finite R M :=
+  ⟨fun _ ↦ inferInstance, fun _ ↦ finitePresentation_of_finite R M⟩
+
+variable {R M N}
+
+lemma Module.finitePresentation_of_free_of_surjective [Module.Free R M] [Module.Finite R M]
+    (l : M →ₗ[R] N)
+    (hl : Function.Surjective l) (hl' : Submodule.FG (LinearMap.ker l)) :
+    Module.FinitePresentation R N := by
+  classical
```
Comment (kbuzzard) [inline Mathlib/Algebra/Module/FinitePresentation.lean:79]: Shouldn't code be indented two further spaces after `classical`? Looking through mathlib it seems that the vast majority of `classical`s have this property. Also on line 122.

File: Mathlib/Algebra/Module/FinitePresentation.lean [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+  · intro hx
+    refine ⟨b.repr.symm (x.mapDomain σ), ?_, ?_⟩
+    · simp [Finsupp.apply_total, hσ₂, hx]
+    · simp only [f, LinearMap.comp_apply, b.repr.apply_symm_apply,
+        LinearEquiv.coe_toLinearMap, Finsupp.lmapDomain_apply]
+      rw [← Finsupp.mapDomain_comp, hσ₁, Finsupp.mapDomain_id]
+  · rintro ⟨y, hy, rfl⟩
+    simp [f, hπ, ← Finsupp.apply_total, hy]
+
+-- Ideally this should be an instance but it makes mathlib much slower.
```
Comment (kbuzzard) [inline Mathlib/Algebra/Module/FinitePresentation.lean:106]: Even with e.g. priority 90?

File: Mathlib/Algebra/Module/FinitePresentation.lean [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+    Submodule.FG (LinearMap.ker l) := by
+  classical
+  obtain ⟨s, hs, hs'⟩ := h
+  have H : Function.Surjective (Finsupp.total s N R Subtype.val) :=
+    LinearMap.range_eq_top.mp (by rw [Finsupp.range_total, Subtype.range_val, ← hs]; rfl)
+  obtain ⟨f, hf⟩ : ∃ f : (s →₀ R) →ₗ[R] M, l ∘ₗ f = (Finsupp.total s N R Subtype.val) := by
+    choose f hf using show _ from hl
+    exact ⟨Finsupp.total s M R (fun i ↦ f i), by ext; simp [hf]⟩
+  apply Submodule.fg_of_fg_map_of_fg_inf_ker f.range.mkQ
+  · convert Module.Finite.out
```
Comment (kbuzzard) [inline Mathlib/Algebra/Module/FinitePresentation.lean:147]: `convert` followed by a bunch of tactics is sometimes a style which is hard to maintain. Is there a cheap way of avoiding the `convert`? If not then don't worry.

**Review by kbuzzard — COMMENTED** [review]
This looks pretty good to me. I left some comments.

**Review by riccardobrasca — APPROVED** [review]

File: Mathlib/Algebra/Module/FinitePresentation.lean [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+  letI : AddCommGroup (LinearMap.ker (l ∘ₗ π)) := inferInstance
+  let f : LinearMap.ker (l ∘ₗ π) →ₗ[R] LinearMap.ker l := LinearMap.restrict π (fun x ↦ id)
+  have e : π ∘ₗ Submodule.subtype _ = Submodule.subtype _ ∘ₗ f := by ext; rfl
+  have hf : Function.Surjective f := by
+    rw [← LinearMap.range_eq_top]
+    apply Submodule.map_injective_of_injective (Submodule.injective_subtype _)
+    rw [Submodule.map_top, Submodule.range_subtype, ← LinearMap.range_comp, ← e,
+      LinearMap.range_comp, Submodule.range_subtype, LinearMap.ker_comp,
+      Submodule.map_comap_eq_of_surjective H]
+  show (LinearMap.ker π).FG
```
[the remark this answers is not in the document]
Comment (erdOne) [reply Mathlib/Algebra/Module/FinitePresentation.lean:186]: No. The goal is `(LinearMap.ker blah).FG` and there is `let π := blah` a few lines prior.

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**